docs/configuration.md

```diff
@@ -112,14 +114,15 @@ spec:
-| ------------------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
-| `externalAddress`        | The loadbalancer address (for k0s controllers running behind a loadbalancer). Configures all cluster components to connect to this address and also configures this address for use  when joining new nodes to the cluster. |
-| `address`                | Local address on which to bind an API. Also serves as one of the addresses pushed on the k0s create service certificate on the API. Defaults to first non-local address found on the node.                                  |
-| `sans`                   | List of additional addresses to push to API servers serving the certificate.                                                                                                                                                |
-| `extraArgs`              | Map of key-values (strings) for any extra arguments to pass down to Kubernetes api-server process.                                                                                                                          |
-| `port`¹                  | Custom port for kube-api server to listen on (default: 6443)                                                                                                                                                                |
-| `k0sApiPort`¹            | Custom port for k0s-api server to listen on (default: 9443)                                                                                                                                                                 |
+| Element             | Description                                                                                                                                                                                                                  |
+|---------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `address`           | Local address on which to bind an API. Also serves as one of the addresses pushed on the k0s create service certificate on the API. Defaults to first non-local address found on the node.                                   |
+| `onlyBindToAddress` | Whether to bind only to the IP address given by the `address` option, instead of binding to all addresses.                                                                                                                   |
```

Comment:
I wonder if we could just keep `bindAddress` instead of `onlyBindToAddress` and let the user decide which address to bind to instead of forcing it to be either the advertised address or all interfaces. This may be a niche requirement, though. I cannot really judge. Maybe some folks with more network expertise have an opinion? OTOH, the `address` field will be defaulted by k0s's address discovery if omitted. There wouldn't be an easy way to let users use the auto-detected address as the bind address, then.

Anyway, I'm fine with the `onlyBindToAddress` logic as well, but wanted to at least ask about it.

Reply:
IMO `bindAddress` makes more sense and is probably more common in general. At least I remember seeing such knobs on other things quite often. 😄

Reply:
Agreed. However, what about

> There wouldn't be an easy way to let users use the auto-detected address as the bind address, then.

Reply:
hmm, what's wrong in the current default behaviour of autodetecting the publish address and binding to all interfaces?

Reply:
> IMO bindAddress makes more sense and is probably more common in general. At least I remember seeing such knobs on other things quite often.

bind-address is a common term yes, but rarely any software asks you to correctly configure several addresses for different purposes, that may or may not interact with each other. That's why I intended to reduce the bind-address parameter to what makes sense. I think the current documentation of `address` is misleading, given nothing actually binds to that address, it's just the internal address published to other components.

> hmm, what's wrong in the current default behaviour of autodetecting the publish address and binding to all interfaces?

The original issue #957 describe where this PR comes from: Multi-homed kubernetes hosts. In my case specially, my k0s hosts was also my edge router, so one of the interfaces the API server bound to was my internet uplink. With proper firewall rules in place no big deal, but it carries some risk. My setup since changed (for the better).

> There wouldn't be an easy way to let users use the auto-detected address as the bind address, then.

if the user configured `bindToAddressOnly: true` without actually configuring the `address`, then it would just use the auto-detected address, right? (probably worth documenting too)

As I said in https://github.com/k0sproject/k0s/pull/3824#discussion_r1622862280, `address` and `bindAddress` are strongly related. Configuring them incorrectly will break the cluster. You can't bind to an address that's not internally announced (unless NAT between components is supported, which would be ridiculous), so you have 2 options: bind to all interfaces and announce one of the addresses or bind to the address you announce. I can't see any other valid configuration (again: assuming NAT is not a thing here). 

Also... I'd assume (I don't have numbers on it) fairly few people actually have a need for this. Few kubernetes hosts are multi-homed and of those most are fine with binding all interfaces.